From b67184480544392f391d73253807b629b614a379 Mon Sep 17 00:00:00 2001
From: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Date: Mon, 1 Jul 2024 13:54:30 +0200
Subject: [PATCH 1007/1695] framework: add prefix when printing a log trace.

Add a prefix for each log level of the messages printed on the console.
The print of the prefix is disabled by default.
It can be activated with the CMake variable : SCP_ENABLE_LOG_PREFIX

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 00f41814ffd188ac42973e7283e217603be5712e framework: add prefix when printing a log trace.]
Change-Id: I25fc43d4a7d47edaca24888d7f3e144c143e03df
Signed-off-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/391889
---
 .../SCP-firmware/framework/CMakeLists.txt     |  4 +++
 .../SCP-firmware/framework/include/fwk_log.h  | 32 +++++++++++++++----
 .../SCP-firmware/framework/src/fwk_log.c      | 19 ++++++++---
 3 files changed, 44 insertions(+), 11 deletions(-)

diff --git a/core/lib/scmi-server/SCP-firmware/framework/CMakeLists.txt b/core/lib/scmi-server/SCP-firmware/framework/CMakeLists.txt
index 60384f39c..1ce4b80d8 100644
--- a/core/lib/scmi-server/SCP-firmware/framework/CMakeLists.txt
+++ b/core/lib/scmi-server/SCP-firmware/framework/CMakeLists.txt
@@ -126,6 +126,10 @@ endif()
 target_compile_definitions(framework
                            PUBLIC FWK_LOG_LEVEL=FWK_LOG_LEVEL_${SCP_LOG_LEVEL})
 
+if(SCP_ENABLE_LOG_PREFIX)
+    target_compile_definitions(framework PUBLIC "FWK_LOG_LEVEL_PREFIX")
+endif()
+
 #
 # Generate `<fwk_module_idx.h>` and `<fwk_module_list.c>` based on the list of
 # modules we have been given. While we're at it, we add these modules as
diff --git a/core/lib/scmi-server/SCP-firmware/framework/include/fwk_log.h b/core/lib/scmi-server/SCP-firmware/framework/include/fwk_log.h
index c13f16d88..f825d22d7 100644
--- a/core/lib/scmi-server/SCP-firmware/framework/include/fwk_log.h
+++ b/core/lib/scmi-server/SCP-firmware/framework/include/fwk_log.h
@@ -200,6 +200,22 @@
 #define FWK_LOG_LEVEL_CRIT 4
 #define FWK_LOG_LEVEL_DISABLED 5
 
+#ifdef FWK_LOG_LEVEL_PREFIX
+#define FWK_PREFIX_TRACE "[TRACE] "
+#define FWK_LOG_PREFIX_DEBUG "[DEBUG] "
+#define FWK_LOG_PREFIX_INFO "[INFO] "
+#define FWK_LOG_PREFIX_WARN "[WARN] "
+#define FWK_LOG_PREFIX_ERR "[ERR] "
+#define FWK_LOG_PREFIX_CRIT "[CRIT] "
+#else
+#define FWK_PREFIX_TRACE NULL
+#define FWK_LOG_PREFIX_DEBUG NULL
+#define FWK_LOG_PREFIX_INFO NULL
+#define FWK_LOG_PREFIX_WARN NULL
+#define FWK_LOG_PREFIX_ERR NULL
+#define FWK_LOG_PREFIX_CRIT NULL
+#endif
+
 /*!
  * \}
  */
@@ -226,7 +242,7 @@
  */
 
 #ifdef FWK_TRACE_ENABLE
-#    define FWK_TRACE(...) fwk_log_printf(__VA_ARGS__)
+#    define FWK_TRACE(...) fwk_log_printf(FWK_PREFIX_TRACE, __VA_ARGS__)
 #else
 #    define FWK_TRACE(...)
 #endif
@@ -240,7 +256,7 @@
  */
 
 #if FWK_LOG_LEVEL <= FWK_LOG_LEVEL_DEBUG
-#    define FWK_LOG_DEBUG(...) fwk_log_printf(__VA_ARGS__)
+#    define FWK_LOG_DEBUG(...) fwk_log_printf(FWK_LOG_PREFIX_DEBUG, __VA_ARGS__)
 #else
 #    define FWK_LOG_DEBUG(...)
 #endif
@@ -254,7 +270,7 @@
  */
 
 #if FWK_LOG_LEVEL <= FWK_LOG_LEVEL_INFO
-#    define FWK_LOG_INFO(...) fwk_log_printf(__VA_ARGS__)
+#    define FWK_LOG_INFO(...) fwk_log_printf(FWK_LOG_PREFIX_INFO, __VA_ARGS__)
 #else
 #    define FWK_LOG_INFO(...)
 #endif
@@ -268,7 +284,7 @@
  */
 
 #if FWK_LOG_LEVEL <= FWK_LOG_LEVEL_WARN
-#    define FWK_LOG_WARN(...) fwk_log_printf(__VA_ARGS__)
+#    define FWK_LOG_WARN(...) fwk_log_printf(FWK_LOG_PREFIX_WARN, __VA_ARGS__)
 #else
 #    define FWK_LOG_WARN(...)
 #endif
@@ -282,7 +298,7 @@
  */
 
 #if FWK_LOG_LEVEL <= FWK_LOG_LEVEL_ERROR
-#    define FWK_LOG_ERR(...) fwk_log_printf(__VA_ARGS__)
+#    define FWK_LOG_ERR(...) fwk_log_printf(FWK_LOG_PREFIX_ERR, __VA_ARGS__)
 #else
 #    define FWK_LOG_ERR(...)
 #endif
@@ -296,7 +312,7 @@
  */
 
 #if FWK_LOG_LEVEL <= FWK_LOG_LEVEL_CRIT
-#    define FWK_LOG_CRIT(...) fwk_log_printf(__VA_ARGS__)
+#    define FWK_LOG_CRIT(...) fwk_log_printf(FWK_LOG_PREFIX_CRIT, __VA_ARGS__)
 #else
 #    define FWK_LOG_CRIT(...)
 #endif
@@ -306,10 +322,12 @@
  *
  * \brief Log a message with a specified filter level.
  *
+ * \param[in] level_string Log level identifier.
  * \param[in] format Format string.
  * \param[in] ... Associated parameters.
  */
-void fwk_log_printf(const char *format, ...) FWK_PRINTF(1, 2);
+void fwk_log_printf(const char* level_string,
+                    const char *format, ...) FWK_PRINTF(2, 3);
 
 /*!
  * \internal
diff --git a/core/lib/scmi-server/SCP-firmware/framework/src/fwk_log.c b/core/lib/scmi-server/SCP-firmware/framework/src/fwk_log.c
index 3be86b37f..34ebc95d2 100644
--- a/core/lib/scmi-server/SCP-firmware/framework/src/fwk_log.c
+++ b/core/lib/scmi-server/SCP-firmware/framework/src/fwk_log.c
@@ -86,6 +86,7 @@ static bool fwk_log_buffer(struct fwk_ring *ring, const char *message)
 #endif
 
 static void fwk_log_vsnprintf(
+    const char *level_string,
     size_t buffer_size,
     char buffer[buffer_size],
     const char *format,
@@ -131,6 +132,16 @@ static void fwk_log_vsnprintf(
         duration_us);
     fwk_assert(length < buffer_size);
 
+    /* Add level string indicator */
+    if (level_string){
+        length += (size_t)snprintf(
+            buffer + length,
+            buffer_size - length,
+            "%s",
+            level_string);
+        fwk_assert(length < buffer_size);
+    }
+
     /*
      * We then need to `snprintf()` the message into a temporary buffer because
      * we need to manipulate it before we print or store it.
@@ -157,7 +168,7 @@ static void fwk_log_snprintf(
     va_list args;
 
     va_start(args, format);
-    fwk_log_vsnprintf(buffer_size, buffer, format, &args);
+    fwk_log_vsnprintf(NULL, buffer_size, buffer, format, &args);
     va_end(args);
 }
 
@@ -189,7 +200,7 @@ static bool fwk_log_banner(void)
     return true;
 }
 
-void fwk_log_printf(const char *format, ...)
+void fwk_log_printf(const char *level_string, const char *format, ...)
 {
     unsigned int flags;
     static bool banner = false;
@@ -212,7 +223,7 @@ void fwk_log_printf(const char *format, ...)
     }
 
     va_start(args, format);
-    fwk_log_vsnprintf(sizeof(buffer), buffer, format, &args);
+    fwk_log_vsnprintf(level_string, sizeof(buffer), buffer, format, &args);
     va_end(args);
 
 #ifdef FWK_LOG_BUFFERED
@@ -273,7 +284,7 @@ int fwk_log_unbuffer(void)
 
             if (fwk_log_ctx.dropped > 0) {
                 fwk_log_printf(
-                    "[FWK] ... and %u more messages...", fwk_log_ctx.dropped);
+                    NULL, "[FWK] ... and %u more messages...", fwk_log_ctx.dropped);
 
                 fwk_log_ctx.dropped = 0;
 
-- 
2.47.2

